### PR TITLE
Valid string units and standard name

### PR DESCRIPTION
**Pull request by benjwadams**
Fixes a number of issues related to non-string units and standard_name attributes in variables.  Addresses #542

### CODE REVIEW

**Comment by benjwadams** [comment]
This also prevents compliance checker from crashing outright if there are variables with non-string standard name.

@kwilcox, feel free to comment on implementation if you'd like.


Aside from unit tests updated here, also tested against the following .cdl:

```

netcdf bad_units_standard_name {
dimensions:
        n_hgrid_face = 2 ;
variables: 
    float hgrid_face_x(n_hgrid_face) ;
        hgrid_face_x:long_name = "x_coordinate of 2D mesh face" ;
        hgrid_face_x:standard_name = 0 ;
        hgrid_face_x:units = "m" ;
        hgrid_face_x:mesh = "hgrid" ;
    float test_var(n_hgrid_face) ;
        test_var:long_name = "Dummy variable for practical salinity" ;
        test_var:standard_name = "sea_water_practical_salinity" ;
        test_var:units = 1 ;
        test_var:mesh = "hgrid" ;
}
```

File: compliance_checker/cfutil.py [hunk 1]
```diff
@@ -52,6 +54,46 @@
     'atmosphere_sleve_coordinate'
 ]
 
+def check_attr_membership(attr_val, value_set, attr_type=basestring,
```
Comment (Bobfrat) [inline compliance_checker/cfutil.py]: If this is just a helper function, I'd suggest changing the name away from the ```check_``` convention we've reserved for checks in the check classes.

Comment (benjwadams) [reply compliance_checker/cfutil.py]: Good point, will fix.

File: compliance_checker/cf/cf.py [hunk 2]
```diff
@@ -895,9 +895,14 @@ def check_units(self, ds):
             valid_units = self._check_valid_cf_units(ds, name)
             ret_val.append(valid_units)
 
-            if isinstance(units, basestring):
+            units_attr_is_string = TestCtx(BaseCheck.MEDIUM,
+                                "§3.1 Variable {}'s units attribute is a string".format(variable.name))
+            # side effects, but better than teasing out the individual result
+            if units_attr_is_string.assert_true(isinstance(units, basestring),
+                                                "'units' attribute must be a string"):
```
[the remark this answers is not in the document]
Comment (benjwadams) [reply compliance_checker/cf/cf.py]: Both should be checked, so I'm going to remove the conditional here and run both checks.

[the remark this answers is not in the document]
Comment (benjwadams) [reply compliance_checker/cf/cf.py]: integer value of `1` is valid udunits according to cfunits:

```python
from cfunits import Unit
# does not throw exception
Unit(1)
```

File: compliance_checker/cfutil.py [hunk 3]
```diff
@@ -52,12 +54,55 @@
...
+        is_in_set = modifier_fn(new_attr_val) in value_set
+    except Exception as e:
+        warnings.warn('Could not apply modifier function {} to value: '
+                      ' {}'.format(modifier_fn, e.msg))
+        return False
+
+    return is_in_set
+
 
 def get_unitless_standard_names(xml_tree, units):
```
Comment (Bobfrat) [inline compliance_checker/cfutil.py]: This is confusing, above in cf.py we're passing in standard_name as the second arg.

```
std_name_unitless = cfutil.get_unitless_standard_names(self._std_names._root, standard_name)
```

Comment (benjwadams) [reply compliance_checker/cfutil.py]: Given that the above is going to be implemented in https://github.com/ioos/compliance-checker/pull/597 instead, please approve this PR so we can merge and hopefully close out #542 or indicate where you may want further improvements.

**Pull request merged**